lib/ipam/ipam.go

```diff
@@ -548,7 +549,44 @@ func (s *blockAssignState) findOrClaimBlock(ctx context.Context, minFreeIps int)
 	return nil, false, errors.New("failed to find or claim a block")
 }
 
-func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, attrs map[string]string, requestedPools []net.IPNet, version int, host string, maxNumBlocks int, rsvdAttr *HostReservedAttr) ([]net.IPNet, error) {
+type IPAMAssignments struct {
+	IPs              []net.IPNet       // assigned IP addresses
+	IPVersion        int               // IP version (4 or 6)
+	NumRequested     int               // number of requested IP addresses (not all may be assigned)
+	HostReservedAttr *HostReservedAttr // reserved addresses at start and/or end of blocks
+	Msgs             []string          // warning/error messages to be rendered in case there are any issues with the assignment
+}
+
+func (i *IPAMAssignments) AddMsg(msg string) {
+	for _, m := range i.Msgs {
+		if msg == m {
+			// Don't add duplicate msgs
+			return
+		}
+	}
+	i.Msgs = append(i.Msgs, msg)
+}
+
+func (i *IPAMAssignments) PartialFulfillmentError() error {
+	if len(i.IPs) < i.NumRequested {
+		var b strings.Builder
+
+		fmt.Fprintf(&b, "Assigned %d out of %d requested IPv%d addresses", len(i.IPs), i.NumRequested, i.IPVersion)
+
+		if len(i.Msgs) > 0 {
+			fmt.Fprintf(&b, "; %v", strings.Join(i.Msgs, ";"))
```

Comment:
I'm also wondering if we're only using `Msgs` to just keep a list of additional info if we should even store it as separate strings or if we should just keep it as a string builder and then just `WriteString` the new messages to it. As far as I can tell we only want to keep them separate so we can do the comparison to see if we've already seen the error message before right?

Reply:
do we need to do the join with `"; "` instead of `";"`?

Reply:
Good catch, thanks! Will fix